feat: add `lapack/base/dlascl` (#7515)
stdlib-js/stdlib: lib/node_modules/@stdlib/lapack/base/dlascl/README.md

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dlascl

> > LAPACK routine to multiply a real M by N matrix `A` by a real scalar `CTO/CFROM`.

Reviewer: ```suggestion
> LAPACK routine to multiply a double-precision floating-point M-by-N matrix `A` by a double-precision floating-point scalar.
```

> 

Reviewer: I suggest adding an intro section stating something like:

This routine multiplies each element of an `M`-by-`N` matrix `A` by a scalar `\alpha`

```math
B = \alpha \cdot A = \alpha \cdot \begin{bmatrix} a_{11} a_{12} \ldots a_{0N} && ...  \end{bmatrix} =  \begin{bmatrix} \alpha a_{11} \alpha a_{12} \ldots \alpha a_{0N} && ...  \end{bmatrix}
```

In order to more readily avoid underflow/overflow, this routine chooses to represent `\alpha` as the product of

```math
\alpha = \beta \cdot \frac{1}{\gamma}
```

<section class="usage">

## Usage

```javascript
var dlascl = require( '@stdlib/lapack/base/dlascl' );
```

> #### dlascl( order, type, KL, KU, CFROM, CTO, M, N, A, LDA )

Reviewer: ```suggestion
#### dlascl( order, type, KL, KU, \gamma, \beta, M, N, A, LDA )
```

Reviewer: Use the actual greek characters.

Reviewer: In JSDoc, use `gamma` and `beta`.

Multiplies a real M by N matrix `A` by a real scalar `CTO/CFROM`.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] ); // => [ [ 1.0, 2.0 ], [ 3.0, 4.0 ], [ 5.0, 6.0 ] ]

dlascl( 'row-major', 'general', 0, 0, 1.0, 2.0, 3, 2, A, 2 );
// A => <Float64Array>[ 2.0, 4.0, 6.0, 8.0, 10.0, 12.0 ]
```

The function has the following parameters:

-   **order**: storage layout.

> -   **type**: specifies the type of matrix `A` ( should be one of these: `general`, `upper`, `lower`, `upper-hessenberg`, `symmetric-banded-lower`, `symmetric-banded-upper` or `banded` ).

Reviewer: ```suggestion
-   **type**: specifies the type of matrix `A`. Must be one of the following: `'general'`, `'upper'`, `'lower'`, `'upper-hessenberg'`, `'symmetric-banded-lower'`, `'symmetric-banded-upper'`, or `'banded'`.
```

> -   **KL**: lower band width of `A`. Referenced only if type is `symmetric-banded-lower` or `banded`.

Reviewer: ```suggestion
-   **KL**: lower bandwidth of `A` (i.e., the number of sub-diagonals). Referenced only if type is `symmetric-banded-lower` or `banded`.
```

> -   **KU**: upper band width of `A`. Referenced only if type is `symmetric-banded-upper` or `banded`.

Reviewer: ```suggestion
-   **KU**: upper bandwidth of `A` (i.e., the number of super-diagonals). Referenced only if type is `symmetric-banded-upper` or `banded`.
```

> -   **CFROM**: the matrix `A` is multiplied by `CTO / CFROM`.

Reviewer: ```suggestion
-   **gamma**: the matrix `A` is multiplied by `beta/gamma`.
```

-   **CTO**: the matrix `A` is multiplied by `CTO / CFROM`.
-   **M**: number of rows in matrix `A`.
-   **N**: number of columns in matrix `A`.
-   **A**: input [`Float64Array`][mdn-float64array].
-   **LDA**: stride of the first dimension of `A` (a.k.a., leading dimension of the matrix `A`).

> If `type` is `banded`, `symmetric-banded-lower` or `symmetric-banded-upper` the matrix should be stored in the [`Band storage`][lapack-band-storage] format.
> 

Reviewer: ```suggestion
```

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

<!-- eslint-disable stdlib/capitalized-comments -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

> // Initial arrays...
> var A0 = new Float64Array( [ 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
> 
> // Create offset views...

Reviewer: ```suggestion
// Initial array:
var A0 = new Float64Array( [ 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );

// Create offset view:
```

var A1 = new Float64Array( A0.buffer, A0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

dlascl( 'row-major', 'general', 0, 0, 1.0, 2.0, 3, 2, A1, 2 );
// A0 => <Float64Array>[ 0.0, 2.0, 4.0, 6.0, 8.0, 10.0, 12.0 ]
```

#### dlascl.ndarray( type, KL, KU, CFROM, CTO, M, N, A, strideA1, strideA2, offsetA )

Multiplies a real M by N matrix `A` by a real scalar `CTO/CFROM` using alternative indexing semantics.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] ); // => [ [ 1.0, 2.0 ], [ 3.0, 4.0 ], [ 5.0, 6.0 ] ]

dlascl.ndarray( 'general', 0, 0, 1.0, 2.0, 3, 2, A, 2, 1, 0 );
// A => <Float64Array>[ 2.0, 4.0, 6.0, 8.0, 10.0, 12.0 ]
```

> The function has the following additional parameters:

Reviewer: ```suggestion
The function has the following parameters:
```

-   **type**: specifies the type of matrix `A` ( should be one of these: `general`, `upper`, `lower`, `upper-hessenberg`, `symmetric-banded-lower`, `symmetric-banded-upper` or `banded` ).
-   **KL**: lower band width of `A`. Referenced only if type is `symmetric-banded-lower` or `banded`.
-   **KU**: upper band width of `A`. Referenced only if type is `symmetric-banded-upper` or `banded`.
-   **CFROM**: the matrix `A` is multiplied by `CTO / CFROM`.
-   **CTO**: the matrix `A` is multiplied by `CTO / CFROM`.
-   **M**: number of rows in matrix `A`.
-   **N**: number of columns in matrix `A`.
-   **A**: input [`Float64Array`][mdn-float64array].
-   **strideA1**: stride of the first dimension of `A`.
-   **strideA2**: stride of the second dimension of `A`.
-   **offsetA**: starting index for `A`.

If `type` is `banded`, `symmetric-banded-lower` or `symmetric-banded-upper` the matrix should be stored in the [`Band storage`][lapack-band-storage] format.

> 

Reviewer: ```suggestion
```

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example,

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var A = new Float64Array( [ 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );

dlascl.ndarray( 'general', 0, 0, 1.0, 2.0, 3, 2, A, 2, 1, 1 );
// A => <Float64Array>[ 0.0, 2.0, 4.0, 6.0, 8.0, 10.0, 12.0 ]
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   `dlascl()` corresponds to the [LAPACK][lapack] routine [`dlascl`][lapack-dlascl].

> -   If `type` is `banded`, `symmetric-banded-lower` or `symmetric-banded-upper` the matrix should be stored in the [`Band storage`][lapack-band-storage] format.

Reviewer: ```suggestion
-   If `type` is `'banded'`, `'symmetric-banded-lower'`, or `'symmetric-banded-upper'`, the matrix should be stored in the [band storage][lapack-band-storage] format.
```

Reviewer: May be good to provide a brief schematic using math equations in terms of what is meant.

Reviewer: @aayush0325 Were you thinking you wanted to do the schematics in a separate document or would it make sense to show the matrix equations in this README (e.g., in the intro)?
